File: nim/src/frq_core.nim
## The C ABI, and nothing else.
##
## Every exported symbol is here so there is one file to read when asking what
## the Dart side can call. The logic lives under `frq/` in ordinary Nim with
## ordinary Nim types, which is what lets the tests test the rules rather than
## the marshalling.
##
## Two conventions, both of which the bindings in `flutter/src/frq/core/`
## wrap so no call site has to remember them:
##
## * Every returned string is the **caller's** to free, with `frq_free`. Nim's
##   allocator is not Dart's, so a `free()` on this side of the boundary is
##   undefined behaviour rather than a leak you can live with.
## * `frq_init` runs once before anything else. Nim's runtime needs setting up
##   and `--app:lib` does not do it for you on every platform.
##
## Answers that are not a single string come back as JSON. A struct would mean
## both sides agreeing on a memory layout, and a field added later would be a
## version skew that segfaults rather than one that fails; JSON costs a parse
## per call, which is nothing against the network round trip that produced the
## line being parsed.
##
## The UI half of this ABI — `frq_ui_render` and `frq_ui_dispatch` — is Nim
## owning the screens as well as the rules. The screens under it are ported
## from `common/frq/screens/` rather than reimagined, which is the difference
## between this and the experiment that was deleted for being a facsimile.

import std/[json, strutils, tables]
import frq/[ircparse, trace, ui, cells, reducer, model, rooms, eintr]
import frq/conn as tr
import frq/screens/connect as scConnectScreen
import frq/screens/chats as scChatsScreen
import frq/screens/chat as scChatScreen
import frq/screens/settings as scSettingsScreen

proc frq_init*() {.exportc, dynlib.} =
  ## Reads the saved sign-in, and otherwise stays out of the way.
  ##
  ## It used to call `NimMain()`. On Linux `--app:lib` already emits a library
  ## constructor that runs Nim's module initialisers at dlopen, so calling it
  ## again ran every module's top-level code a SECOND time — which for
  ## `conn.nim` meant `outbound.open()` on channels that were already open,
  ## quietly resetting them. The reader thread then drained a different queue
  ## from the one the writer filled, and nothing this client sent ever left.
  ##
  ## So it stayed empty for a long time. What it does now is the one thing
  ## that genuinely belongs before the first frame and cannot go in
  ## `initState`, which is a `func` and touches no disk: restoring the broker
  ## token, so the connect screen opens saying the session is remembered
  ## rather than offering a login page the reader does not need.
  ##
  ## `frq_ui_reset` deliberately does not do this. It is the tests' entry
  ## point, and a suite that picked up whoever is signed in on the machine
  ## running it would pass or fail by accident.
  # Before anything opens a socket. The Dart VM's profiler signals every
  # thread in this process about a thousand times a second, and a syscall
  # interrupted by one fails rather than resuming unless its handler says
  # otherwise — which is why signing in reported `Interrupted system call`.
  restartableSyscalls()
  reducer.restore()

proc dup(s: string): cstring =
  ## A copy of `s` that outlives this call, for the caller to `frq_free`.
  ## `allocShared0` and not `alloc0`: the Dart side may free it from a
  ## different thread than the one that made it.
  let n = s.len
  let p = cast[cstring](allocShared0(n + 1))
  if n > 0:
    copyMem(p, unsafeAddr s[0], n)
  p

proc frq_free*(p: cstring) {.exportc, dynlib.} =
  ## Free what one of the functions below returned. Null is fine.
  if p != nil:
    deallocShared(p)

proc frq_version*(): cstring {.exportc, dynlib.} =
  ## Static storage, deliberately: this one is NOT freed, and is the only
  ## exception to the rule above. It exists so a binding can check at load
  ## time that the library it found is the one it was built against.
  "0.1.0"

# ----------------------------------------------------------------- irc/parse

proc frq_irc_parse_line*(line: cstring): cstring {.exportc, dynlib.} =
  ## An IRC line as JSON: `{raw, tags, account, prefix, command, params}`.
  ##
  ## `tags`, `account` and `prefix` are JSON null where the line carried none,
  ## which is the distinction `frq.irc.parse` draws with nil and every caller
  ## of it depends on — a PRIVMSG from a server with no prefix is not the same
  ## line as one from a nick.
  if line == nil: return dup("null")
  let p = parseLine($line)
  var o = newJObject()
  o["raw"] = %p.raw
  o["tags"] = if p.hasTags: %p.tags else: newJNull()
  o["account"] = if p.hasAccount: %p.account else: newJNull()
  o["prefix"] = if p.hasPrefix: %p.prefix else: newJNull()
  o["command"] = %p.command
  o["params"] = %p.params
  dup($o)

proc frq_irc_tag_value*(tags, key: cstring): cstring {.exportc, dynlib.} =
  ## One tag's value, unescaped — or **null** where the tag is absent or
  ## empty, which IRCv3 says are the same thing. Null and not "" on purpose:
  ## see `tagValue`.
  if tags == nil or key == nil: return nil
  let (v, ok) = tagValue($tags, $key)
  if ok: dup(v) else: nil

proc frq_irc_unescape_tag*(v: cstring): cstring {.exportc, dynlib.} =
  if v == nil: return nil
  dup(unescapeTag($v))

proc frq_irc_escape_tag_value*(v: cstring): cstring {.exportc, dynlib.} =
  if v == nil: return dup("")
  dup(escapeTagValue($v))

proc frq_irc_nick_of*(prefix: cstring): cstring {.exportc, dynlib.} =
  if prefix == nil: return nil
  dup(nickOf($prefix))

# --------------------------------------------------------------- transport
#
# `frq.net`'s three operations, for `frq.net.nim` to install. This is the
# wiring that matters: the existing ClojureDart screens, cells and actions are
# untouched, and only the socket underneath them becomes Nim.
#
# Polled rather than callback-driven, for the reason the UI is: a Dart callback
# invoked from a foreign thread has to be marshalled onto the main isolate, and
# a timer on the Dart side does the same job with no mechanism at all.

proc frq_trace*(topic, msg: cstring) {.exportc, dynlib.} =
  ## Let the Dart side log through the same facility, so one FRQ_TRACE=1 gives
  ## one interleaved story instead of two half-ones in different places.
  if topic != nil and msg != nil:
    trace($topic, $msg)

proc frq_conn_open*(host: cstring, port: cint, tls: cint) {.exportc, dynlib.} =
  if host == nil: return
  tr.open(tr.ConnConfig(host: $host, port: port.int, tls: tls != 0))

proc frq_conn_send*(line: cstring) {.exportc, dynlib.} =
  if line != nil: tr.send($line)

proc frq_conn_close*() {.exportc, dynlib.} =
  tr.close()

proc frq_conn_recv*(): cstring {.exportc, dynlib.} =
  ## The next line, or null when there is none waiting. Never blocks.
  let (ok, line) = tr.tryLine()
  if ok: dup(line) else: nil

proc frq_conn_event*(): cstring {.exportc, dynlib.} =
  ## The next transport event — "open", "close: …", "error: …" — or null.
  let (ok, e) = tr.tryEvent()
  if ok: dup(e) else: nil


# ------------------------------------------------------------------- the UI
#
# Nim owns the state and the screens; Dart owns the pixels. The only things
# crossing are a tree going out and an event id coming back.

proc currentTree(): string =
  ## Whichever screen the state says. `drain` first, so the tree Dart gets is
  ## built after every line that had arrived when it asked — that is the whole
  ## of the polling model, and why there is no callback into Dart.
  drain()
  let connected = app.status.startsWith("Connected")
  let node =
    case app.screen
    of scChat: scChatScreen.chatScreen(app, connected)
    of scChats: scChatsScreen.chatsScreen(app, connected)
    of scDiscover: scSettingsScreen.discoverScreen(app)
    of scSettings: scSettingsScreen.settingsScreen(app, connected, true)
    of scConnect: scConnectScreen.connectScreen(app)
  $node.toJson

proc frq_ui_render*(): cstring {.exportc, dynlib.} =
  ## The current screen as a widget tree, in JSON.
  ##
  ## Not pure: it drains the socket's queue first, so two calls with no
  ## dispatch between can differ when a line arrived in the gap. That is how
  ## the room fills, and it is why the renderer polls.
  dup(currentTree())

proc frq_ui_dispatch*(event: cstring): cstring {.exportc, dynlib.} =
  ## Apply an event and answer with the tree it produced.
  ##
  ## One call rather than dispatch-then-render, and not to save a crossing: it
  ## makes the pair atomic, so there is no window in which Dart could render a
  ## state nothing asked for.
  if event != nil:
    try:
      dispatch(parseJson($event))
    except JsonParsingError:
      discard
  dup(currentTree())

proc frq_ui_poll*(): cstring {.exportc, dynlib.} =
  ## The tree, for a renderer asking because time passed rather than because
  ## anything happened. Same work as render; named for what the caller means.
  dup(currentTree())

proc frq_ui_wanted_picture*(): cstring {.exportc, dynlib.} =
  ## What an upload needs — `{host, did, channel}` — or empty where none is
  ## wanted. Picking a file and posting it are the host's; this is the core
  ## saying who is asking and where to.
  dup(reducer.wantedPicture())

proc frq_ui_demo*() {.exportc, dynlib.} =
  ## Fill a room with a representative conversation, for a test that wants to
  ## lay the chat screen out without a server.
  ##
  ## It exists because the chat screen is the one a script could not reach: a
  ## GUI on Wayland cannot be clicked, so every automated check stopped at the
  ## room list and the biggest screen in the app went out unlaid-out. The
  ## content is chosen to be awkward on purpose — a long unbroken URL, a very
  ## long word, an image, reactions, a reply, a system line, an edited line —
  ## because a layout bug is about what does not fit.
  app = initState()
  app.formNick = "me"
  app.rooms.ensureRoom("#test")
  var r = app.rooms["#test"]
  r.joined = true
  r.users = {"me": "", "alice": "@", "bob": ""}.toTable
  r.topic = "a room"
  let t0 = 1_700_000_000_000'i64
  r.messages = @[
    Message(id: "1", frm: "*", text: "me joined #test", at: t0, system: true),
    Message(id: "2", frm: "alice", text: "hello there", at: t0 + 1000),
    Message(id: "3", frm: "bob",
            text: "see https://example.com/a/very/long/path/that/will/not/wrap/anywhere/at/all?q=1 for more",
            at: t0 + 2000),
    Message(id: "4", frm: "alice",
            text: "Supercalifragilisticexpialidociousssssssssssssssssssssssssssssssssssss",
            at: t0 + 3000),
    Message(id: "5", frm: "me", text: "a picture", at: t0 + 4000,
            imageUrl: "https://example.com/a.png"),
    Message(id: "6", frm: "bob", text: "answering you", at: t0 + 5000,
            replyTo: "5"),
    Message(id: "7", frm: "me", text: "edited line", at: t0 + 6000,
            edited: true,
            reactions: @[Reaction(emoji: "👍", nicks: @["me", "alice"]),
                         Reaction(emoji: "🎉", nicks: @["bob"])]),
    # A different day, so a heading has to land between them.
    Message(id: "8", frm: "alice", text: "next day", at: t0 + 200_000_000)]
  app.rooms["#test"] = r

  # Somewhere else, because the overview is about every room but this one and
  # with a single room there was nothing for it to show. The test that laid
  # the overview out was laying out "Nothing has happened anywhere else."
  #
  # Awkward on purpose, like the rest: a long bot line that has to be cut,
  # and a short one that must not look different for it.
  app.rooms.ensureRoom("#tasks")
  var other = app.rooms["#tasks"]
  other.joined = true
  other.messages = @[
    Message(id: "o1", frm: "freeq-bot", text: "sandbox-01: I am sandbox 1 of 50.",
            at: t0 + 100_000),
    Message(id: "o2", frm: "freeq-bot",
            text: "**Result**: 50/50 announcements collected in ~12s, and the rest of a sentence that has to be cut somewhere sensible.",
            at: t0 + 200_000)]
  app.rooms["#tasks"] = other

  # A third room, and enough in it that the overview is taller than a phone.
  # Without one nothing here would ever need to scroll, and a test that the
  # overview scrolls would pass on a list that fits.
  app.rooms.ensureRoom("#busy")
  var busy = app.rooms["#busy"]
  busy.joined = true
  for i in 1 .. 12:
    busy.messages.add Message(id: "b" & $i, frm: "carol",
                              text: "line " & $i & " of a long day",
                              at: t0 + 300_000 + i.int64 * 1000)
  app.rooms["#busy"] = busy

  app.current = "#test"
  app.screen = scChat
  app.status = "Connected as me"

proc frq_ui_reset*() {.exportc, dynlib.} =
  tr.close()
  app = initState()

File: dart/frq_core/lib/frq_core.dart
/// The Nim core, as Dart functions.
///
/// The logic lives in Nim; this is the shape it takes on this side. Two hosts
/// answer it — `src/host_ffi.dart` through `dart:ffi` on a desktop, and
/// `src/host_js.dart` against the `nim js` build in a browser — and
/// `src/host.dart` is the one line that chooses. Nothing below this comment
/// knows which, which is the point: the widget tree, its decoding, and the
/// change detection that makes polling cheap are the same work whatever
/// produced the JSON.
///
/// **Dart and not ClojureDart, on purpose.** The point of the Nim core is to
/// have less Clojure, so new code on this side of the boundary is written in
/// the language the platform speaks.
///
/// No `package:ffi` either. That package exists mostly for `Utf8`
/// conversions, and doing them against `dart:convert` costs about ten lines
/// and keeps `pubspec.yaml` unchanged.
library;

import 'dart:convert';

import 'src/host.dart' as host;

// ------------------------------------------------------------------ public

/// The core's version, for a caller that wants to check the library it found
/// is the one it was built against.
String get version => host.hostVersion();

/// An IRC line, taken apart: `{raw, tags, account, prefix, command, params}`.
///
/// `tags`, `account` and `prefix` are null where the line carried none, which
/// is the distinction `frq.irc.parse` draws with nil and every caller depends
/// on — a PRIVMSG from a server with no prefix is not the same line as one
/// from a nick.
Map<String, dynamic> parseLine(String line) {
  final json = host.str1('frq_irc_parse_line', line);
  return jsonDecode(json ?? 'null') as Map<String, dynamic>;
}

/// One IRCv3 tag's value, unescaped — null where the tag is absent OR empty,
/// which IRCv3 says are the same thing.
String? tagValue(String tags, String key) {
  return host.tagValueOf(tags, key);
}

String unescapeTag(String v) => host.str1('frq_irc_unescape_tag', v) ?? '';

String escapeTagValue(String v) => host.str1('frq_irc_escape_tag_value', v) ?? '';

/// The nick half of a `nick!user@host` prefix.
String nickOf(String prefix) => host.str1('frq_irc_nick_of', prefix) ?? '';


/// Log through the Nim core's trace facility, so `FRQ_TRACE=1` gives one
/// interleaved story rather than two half-ones in different places.
void trace(String topic, String msg) {
  host.traceTo(topic, msg);
}

// --------------------------------------------------------------- transport
//
// `frq.net`'s three operations, with a Nim socket behind them. This is the
// wiring that leaves the existing ClojureDart screens, cells and actions
// alone: only the transport underneath them is Nim.


/// Dial. Non-blocking: the socket runs on a Nim thread and progress arrives
/// through [connEvent].
void connOpen(String hostname, int port, {bool tls = true}) =>
    host.connOpenAt(hostname, port, tls);


/// Queue a line. The transport adds the CRLF.
void connSend(String line) => host.connSendLine(line);


void connClose() => host.connCloseNow();

/// The next line, or null when none is waiting. Never blocks.
String? connRecv() => host.connRecvLine();

/// The next transport event — `open`, `close: …`, `error: …` — or null.
String? connEvent() => host.connEventNext();


// ---------------------------------------------------------------- the UI
//
// Nim owns the state and the screens; Dart owns the pixels. A tree goes out,
// an event id comes back, and nothing else crosses.
//
// `UiNode` is deliberately a dumb bag — a tag, a props map, children. A class
// per widget would put the tag vocabulary in two places and make every new tag
// a change on both sides; the point is that Nim can grow a screen without this
// file being touched.

/// One node of the widget tree Nim emitted.
class UiNode {
  final String tag;
  final Map<String, dynamic> props;
  final List<UiNode> children;

  const UiNode(this.tag, this.props, this.children);

  factory UiNode.fromJson(Map<String, dynamic> j) => UiNode(
        j['tag'] as String,
        (j['props'] as Map?)?.cast<String, dynamic>() ?? const {},
        ((j['children'] as List?) ?? const [])
            .map((c) => UiNode.fromJson((c as Map).cast<String, dynamic>()))
            .toList(growable: false),
      );

  /// A prop, or [fallback] when it is absent or the wrong shape. Tolerant on
  /// purpose: a renderer should skip a prop it does not understand rather than
  /// fail a whole screen over one.
  T prop<T>(String name, T fallback) {
    final v = props[name];
    return v is T ? v : fallback;
  }

  /// Structural, and that matters: the poll loop compares two trees by this
  /// string to decide whether to rebuild. A summary showing only tags and prop
  /// NAMES would call two screens equal when a message had arrived, and the
  /// room would never appear to fill.
  @override
  String toString() =>
      '<$tag $props ${children.map((c) => c.toString()).join()}>';
}

UiNode _treeFrom(String? json) =>
    UiNode.fromJson(jsonDecode(json ?? _emptyTree) as Map<String, dynamic>);

/// A tree and the JSON it came from.
///
/// The raw string is kept because it is the cheapest possible change
/// detector: Nim already produced it, and comparing two strings is free
/// beside decoding one. The renderer polls ten times a second and the answer
/// is almost always "nothing changed" — doing a `jsonDecode` and two
/// recursive `toString()`s to discover that was most of the idle cost of the
/// app in a busy room.
const _emptyTree = '{"tag":"vbox"}';

class UiFrame {
  final String json;
  final UiNode tree;
  const UiFrame(this.json, this.tree);
}

/// The current screen.
///
/// Not pure: the Nim side drains the socket's queue first, so two calls with
/// no [dispatch] between can differ when a line arrived in the gap. That is how
/// the room fills, and why the renderer polls.
UiNode render() => _treeFrom(
    host.uiRender());

/// The tree, asked for because time passed rather than because anything
/// happened. Same work as [render]; named for what the caller means.
UiNode poll() => _treeFrom(
    host.uiPoll());

/// The current screen, with the JSON it came from. The starting point for
/// [pollIfChanged].
UiFrame renderFrame() {
  final json = host.uiRender() ?? _emptyTree;
  return UiFrame(json, _treeFrom(json));
}

/// The tree, decoded only when it differs from [since] — otherwise null,
/// meaning "the screen you already have is current".
///
/// This is what the renderer polls with. The comparison is the JSON Nim
/// already produced, so an unchanged frame costs one string compare rather
/// than a decode and two recursive `toString()`s.
UiFrame? pollIfChanged(String since) {
  final json = host.uiPoll() ?? _emptyTree;
  if (json == since) return null;
  return UiFrame(json, _treeFrom(json));
}

/// Apply an event and get the tree it produced.
///
/// One call rather than dispatch-then-render, and not to save a crossing: it
/// makes the pair atomic, so there is no window in which Dart could render a
/// state nothing asked for.
UiNode dispatch(String id, [String value = '']) => dispatchFrame(id, value).tree;

/// As [dispatch], but keeping the JSON so the poll loop can compare against
/// it without re-stringifying the tree it just built.
UiFrame dispatchFrame(String id, [String value = '']) {
  final json = host.uiDispatch(jsonEncode({'id': id, 'value': value})) ??
      _emptyTree;
  return UiFrame(json, _treeFrom(json));
}


/// What an upload needs — `{host, did, channel}` as JSON — or empty where
/// nothing is wanted. Taken as it is read: a file dialog opened twice is one
/// the reader has to dismiss twice.
String wantedPicture() => host.wantedPicture() ?? '';

/// Fill a room with a representative conversation, so a test can lay the chat
/// screen out without a server. See the Nim side for why it exists.
void demoUi() => host.uiDemo();

/// Back to a fresh state, for a caller that wants a known starting point.
void resetUi() => host.uiReset();
